Bug: Problem with custom encoders for Vector[Custom]

@kanterov @OlivierBlanvillain  Found an issue encoding (say) `Vector[Food]` in our tests.

Here is a minimal test to reproduce the error

```scala
test("test me ....") {
    def prop[
    A: TypedEncoder,
    B: TypedEncoder,
    C: TypedEncoder,
    D: TypedEncoder](data: Vector[X4[A, B, C, D]]): Prop = {
      val ds = TypedDataset.create(data)
      ds.show().run()
      true
    }
    check(forAll(prop[X2[Int, Int], Int, Boolean, Vector[Food]] _))
  }
```

Running this gives the following runtime error:

```java
RuntimeException was thrown during property evaluation.
  Message: Error while encoding: java.lang.ClassCastException: frameless.Pasta$ cannot be cast to java.lang.Integer
named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b) AS a#20
+- named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b)
   :- a
   :- input[0, frameless.X4, false].a.a
   :  +- input[0, frameless.X4, false].a
   :     +- input[0, frameless.X4, false]
   :- b
   +- input[0, frameless.X4, false].a.b
      +- input[0, frameless.X4, false].a
         +- input[0, frameless.X4, false]

input[0, frameless.X4, false].b AS b#21
+- input[0, frameless.X4, false].b
   +- input[0, frameless.X4, false]

input[0, frameless.X4, false].c AS c#22
+- input[0, frameless.X4, false].c
   +- input[0, frameless.X4, false]

newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData) AS d#23
+- newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData)
   +- input[0, frameless.X4, false].d
      +- input[0, frameless.X4, false]

  Occurred when passed generated values (
    arg0 = Vector(X4(X2(-1,-1694336440),1160431481,true,Vector(Pasta)))
  )
ScalaTestFailureLocation: frameless.CreateTests$$anonfun$2 at (CreateTests.scala:38)
org.scalatest.exceptions.GeneratorDrivenPropertyCheckFailedException: RuntimeException was thrown during property evaluation.
  Message: Error while encoding: java.lang.ClassCastException: frameless.Pasta$ cannot be cast to java.lang.Integer
named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b) AS a#20
+- named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b)
   :- a
   :- input[0, frameless.X4, false].a.a
   :  +- input[0, frameless.X4, false].a
   :     +- input[0, frameless.X4, false]
   :- b
   +- input[0, frameless.X4, false].a.b
      +- input[0, frameless.X4, false].a
         +- input[0, frameless.X4, false]

input[0, frameless.X4, false].b AS b#21
+- input[0, frameless.X4, false].b
   +- input[0, frameless.X4, false]

input[0, frameless.X4, false].c AS c#22
+- input[0, frameless.X4, false].c
   +- input[0, frameless.X4, false]

newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData) AS d#23
+- newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData)
   +- input[0, frameless.X4, false].d
      +- input[0, frameless.X4, false]

  Occurred when passed generated values (
    arg0 = Vector(X4(X2(-1,-1694336440),1160431481,true,Vector(Pasta)))
  )
	at org.scalatest.prop.Checkers$.doCheck(Checkers.scala:428)
	at org.scalatest.prop.Checkers$class.check(Checkers.scala:346)
	at frameless.TypedDatasetSuite.check(TypedDatasetSuite.scala:8)
	at org.scalatest.prop.Checkers$class.check(Checkers.scala:357)
	at frameless.TypedDatasetSuite.check(TypedDatasetSuite.scala:8)
	at frameless.CreateTests$$anonfun$2.apply$mcV$sp(CreateTests.scala:38)
	at frameless.CreateTests$$anonfun$2.apply(CreateTests.scala:22)
	at frameless.CreateTests$$anonfun$2.apply(CreateTests.scala:22)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
	at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
	at org.scalatest.FunSuite.withFixture(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
	at org.scalatest.FunSuite.runTest(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
	at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
	at org.scalatest.Suite$class.run(Suite.scala:1424)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
	at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
	at org.scalatest.FunSuite.run(FunSuite.scala:1555)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
	at org.scalatest.tools.Runner$.run(Runner.scala:883)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.RuntimeException: Error while encoding: java.lang.ClassCastException: frameless.Pasta$ cannot be cast to java.lang.Integer
named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b) AS a#20
+- named_struct(a, input[0, frameless.X4, false].a.a, b, input[0, frameless.X4, false].a.b)
   :- a
   :- input[0, frameless.X4, false].a.a
   :  +- input[0, frameless.X4, false].a
   :     +- input[0, frameless.X4, false]
   :- b
   +- input[0, frameless.X4, false].a.b
      +- input[0, frameless.X4, false].a
         +- input[0, frameless.X4, false]

input[0, frameless.X4, false].b AS b#21
+- input[0, frameless.X4, false].b
   +- input[0, frameless.X4, false]

input[0, frameless.X4, false].c AS c#22
+- input[0, frameless.X4, false].c
   +- input[0, frameless.X4, false]

newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData) AS d#23
+- newInstance(class org.apache.spark.sql.catalyst.util.GenericArrayData)
   +- input[0, frameless.X4, false].d
      +- input[0, frameless.X4, false]

	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:279)
	at org.apache.spark.sql.SparkSession$$anonfun$3.apply(SparkSession.scala:421)
	at org.apache.spark.sql.SparkSession$$anonfun$3.apply(SparkSession.scala:421)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.apache.spark.sql.SparkSession.createDataset(SparkSession.scala:421)
	at org.apache.spark.sql.SQLContext.createDataset(SQLContext.scala:380)
	at frameless.TypedDataset$.create(TypedDataset.scala:471)
	at frameless.CreateTests$$anonfun$2.frameless$CreateTests$$anonfun$$prop$1(CreateTests.scala:28)
	at frameless.CreateTests$$anonfun$2$$anonfun$apply$mcV$sp$1.apply(CreateTests.scala:38)
	at frameless.CreateTests$$anonfun$2$$anonfun$apply$mcV$sp$1.apply(CreateTests.scala:38)
	at scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
	at org.scalacheck.Prop$$anonfun$forAllShrink$1$$anonfun$3.apply(Prop.scala:714)
	at org.scalacheck.Prop$$anonfun$forAllShrink$1$$anonfun$3.apply(Prop.scala:714)
	at org.scalacheck.Prop$.secure(Prop.scala:458)
	at org.scalacheck.Prop$$anonfun$forAllShrink$1.org$scalacheck$Prop$$anonfun$$result$1(Prop.scala:714)
	at org.scalacheck.Prop$$anonfun$forAllShrink$1.apply(Prop.scala:751)
	at org.scalacheck.Prop$$anonfun$forAllShrink$1.apply(Prop.scala:708)
	at org.scalacheck.Prop$$anonfun$apply$5.apply(Prop.scala:293)
	at org.scalacheck.Prop$$anonfun$apply$5.apply(Prop.scala:292)
	at org.scalacheck.PropFromFun.apply(Prop.scala:21)
	at org.scalacheck.Test$.org$scalacheck$Test$$workerFun$1(Test.scala:323)
	at org.scalacheck.Test$$anonfun$2.apply(Test.scala:352)
	at org.scalacheck.Test$$anonfun$2.apply(Test.scala:352)
	at org.scalacheck.Platform$.runWorkers(Platform.scala:40)
	at org.scalacheck.Test$.check(Test.scala:352)
	at org.scalatest.prop.Checkers$.doCheck(Checkers.scala:372)
	... 56 more
Caused by: java.lang.ClassCastException: frameless.Pasta$ cannot be cast to java.lang.Integer
	at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:101)
	at org.apache.spark.sql.catalyst.util.GenericArrayData.getInt(GenericArrayData.scala:62)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply_1$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:276)
	... 88 more
```

--- a/dataset/src/main/scala/frameless/TypedEncoder.scala
+++ b/dataset/src/main/scala/frameless/TypedEncoder.scala
@@ -283,7 +283,8 @@ object TypedEncoder {
     }
 
     def extractorFor(path: Expression): Expression = {
-      if (ScalaReflection.isNativeType(underlying.targetDataType)) {
+      // if source `path` is already native for Spark, no need to `map`
+      if (ScalaReflection.isNativeType(underlying.sourceDataType)) {
         NewInstance(
           classOf[GenericArrayData],
           path :: Nil,

--- a/dataset/src/test/scala/frameless/CollectTests.scala
+++ b/dataset/src/test/scala/frameless/CollectTests.scala
@@ -41,6 +41,10 @@ class CollectTests extends TypedDatasetSuite {
     check(forAll(prop[Vector[Int]] _))
     check(forAll(prop[Option[Int]] _))
     check(forAll(prop[Vector[X2[Int, Int]]] _))
+
+    check(forAll(prop[X1[Vector[Food]]] _))
+    check(forAll(prop[X1[Vector[X1[Food]]]] _))
+    check(forAll(prop[X1[Vector[X1[Int]]]] _))
   }
 }
 

